### PR TITLE
Optimize `array.size()`, `blob.size()` and `text.size()`

### LINKED ISSUE #3861: perf bug: `array.size()` might be allocating a closure, creating unnecessary garbage.
```
let l : [Nat] = [];    
let pre = Prim.rts_heap_size();
ignore l.size();   
Debug.print(Nat.toText(Prim.rts_heap_size()-pre));
```

https://dfinity.slack.com/archives/CPL67E7MX/p1678192911487829

### PR DESCRIPTION
**Pull request by nomeata**
fixes #3861

### CODE REVIEW

File: src/lowering/desugar.ml [hunk 1]
```diff
@@ -193,6 +193,17 @@ and exp' at note = function
     I.PrimE (I.OtherPrim p, exps es)
   | S.CallE ({it=S.AnnotE ({it=S.PrimE p;_},_);_}, _, e) ->
     I.PrimE (I.OtherPrim p, [exp e])
+  (* Optimizing array.size() *)
+  | S.CallE ({it=S.DotE (arr, proj); _}, _, e1)
+      when T.is_array arr.note.S.note_typ && proj.it = "size" ->
+    (thenE (exp e1) (primE (I.OtherPrim "array_len") [exp arr])).it
+  | S.CallE ({it=S.DotE (arr, proj); _}, _, e1)
+      when T.is_prim T.Text arr.note.S.note_typ && proj.it = "size" ->
```
Comment (ggreif) [inline src/lowering/desugar.ml:201]:
```suggestion
      when T.(is_prim Text) arr.note.S.note_typ && proj.it = "size" ->
```
this should work

**Review by ggreif — APPROVED** [review]
This is very nice, thanks!

**Comment by nomeata** [comment]
No relevant improvement in our tests…

File: src/lowering/desugar.ml [hunk 2]
```diff
@@ -193,6 +193,17 @@ and exp' at note = function
...
+    (thenE (exp e1) (primE (I.OtherPrim "text_len") [exp arr])).it
+  | S.CallE ({it=S.DotE (arr, proj); _}, _, e1)
+      when T.is_prim T.Blob arr.note.S.note_typ && proj.it = "size" ->
```
[the remark this answers is not in the document]
Comment (nomeata) [reply src/lowering/desugar.ml:204]:
```suggestion
      when T.(is_prim Blob) arr.note.S.note_typ && proj.it = "size" ->
```

File: src/lowering/desugar.ml [hunk 3]
```diff
@@ -193,6 +193,17 @@ and exp' at note = function
...
+  (* Optimizing array.size() *)
+  | S.CallE ({it=S.DotE (arr, proj); _}, _, e1)
+      when T.is_array arr.note.S.note_typ && proj.it = "size" ->
+    (thenE (exp e1) (primE (I.OtherPrim "array_len") [exp arr])).it
+  | S.CallE ({it=S.DotE (arr, proj); _}, _, e1)
+      when T.(is_prim Text) arr.note.S.note_typ && proj.it = "size" ->
+    (thenE (exp e1) (primE (I.OtherPrim "text_len") [exp arr])).it
+  | S.CallE ({it=S.DotE (arr, proj); _}, _, e1)
+      when T.(is_prim Blob) arr.note.S.note_typ && proj.it = "size" ->
+    (thenE (exp e1) (primE (I.OtherPrim "blob_size") [exp arr])).it
```
Comment (nomeata) [inline src/lowering/desugar.ml:205]:
```suggestion
  (* Optimizing array.size() *)
  | S.CallE ({it=S.DotE (e1, proj); _}, _, e2)
      when T.is_array arr.note.S.note_typ && proj.it = "size" ->
    (thenE (exp e2) (primE (I.OtherPrim "array_len") [exp e1])).it
  | S.CallE ({it=S.DotE (e1, proj); _}, _, e2)
      when T.(is_prim Text) arr.note.S.note_typ && proj.it = "size" ->
    (thenE (exp e2) (primE (I.OtherPrim "text_len") [exp e1])).it
  | S.CallE ({it=S.DotE (e1, proj); _}, _, e2)
      when T.(is_prim Blob) arr.note.S.note_typ && proj.it = "size" ->
    (thenE (exp e2) (primE (I.OtherPrim "blob_size") [exp e1])).it
```

**Comment by nomeata** [comment]
The current code is bad if the argument to `arr.size` isn't a manifest unit tuple, but some expression (it changes the order of evaluation). Maybe we should simply only match on manifest tuples here; it’s an ad-hoc optimization anyways.

**Comment by crusso** [comment]
https://m7sm4-2iaaa-aaaab-qabra-cai.ic0.app/?tag=2221786361 shows that  this should be an improvement

**Pull request merged**